### PR TITLE
Switch Jaeger receiver to internal data model

### PR DESCRIPTION
**Pull request by dmitryax**
Conversion were replicated from existing jaeger -> OC -> internal.

One concern:
We are getting ParentSpanID from Jaeger span links and writing the link to internal Span.Links slice as well. I'm not sure if it's desired behavior. I'd remove parent span link from the translated internal span and keep that information only ParentSpanID field. Keeping the link copy in Span.Links makes it even more confusing because, with the links conversion, we lose RefType information (like Span_Link_PARENT_LINKED_SPAN or Span_Link_TYPE_UNSPECIFIED), OTLP doesn't have that field.
@tigrannajaryan @pjanotti what do you think about it ^

### CODE REVIEW

File: translator/internaldata/oc_to_resource.go [hunk 1]
```diff
@@ -28,22 +28,6 @@ func ocNodeResourceToInternal(ocNode *occommon.Node, ocResource *ocresource.Reso
...
-	// Calculate maximum total number of attributes. It is OK if we are a bit higher than
-	// the exact number since this is only needed for capacity reservation.
-	maxTotalAttrCount := 0
-	if ocNode != nil {
-		maxTotalAttrCount += len(ocNode.Attributes) + specialNodeAttrCount
-	}
-	if ocResource != nil {
-		maxTotalAttrCount += len(ocResource.Labels) + specialResourceAttrCount
-	}
-
```
Comment (bogdandrutu) [inline translator/internaldata/oc_to_resource.go:46]: why this change?

Comment (dmitryax) [reply translator/internaldata/oc_to_resource.go:46]: This is just dead code

File: receiver/jaegerreceiver/trace_receiver_test.go [hunk 2]
```diff
@@ -217,81 +224,58 @@ func TestGRPCReceptionWithTLS(t *testing.T) {
...
+
+	span0 := rs.InstrumentationLibrarySpans().At(0).Spans().At(0)
+	span0.SetSpanID(childSpanID)
+	span0.SetParentSpanID(parentSpanID)
+	span0.SetTraceID(traceID)
+	span0.SetName("DBSearch")
+	span0.SetStartTime(pdata.TimestampUnixNano(uint64(t1.UnixNano())))
+	span0.SetEndTime(pdata.TimestampUnixNano(uint64(t2.UnixNano())))
+	span0.Status().InitEmpty()
+	span0.Status().SetCode(pdata.StatusCode(otlptrace.Status_NotFound))
```
[the remark this answers is not in the document]
Comment (dmitryax) [reply receiver/jaegerreceiver/trace_receiver_test.go:260]: done https://github.com/open-telemetry/opentelemetry-collector/pull/850

File: translator/trace/jaeger/jaegerproto_to_traces.go [hunk 3]
```diff
@@ -0,0 +1,272 @@
+// Copyright 2019, OpenTelemetry Authors
```
[the remark this answers is not in the document]
Comment (dmitryax) [reply translator/trace/jaeger/jaegerproto_to_traces.go]: changed

File: receiver/jaegerreceiver/factory.go [hunk 4]
```diff
@@ -213,14 +214,15 @@ func (f *Factory) CreateTraceReceiver(
...
 
 	// Create the receiver.
-	return New(rCfg.Name(), &config, nextConsumer, logger)
+	return New(rCfg.Name(), &config, nextConsumer, params)
 }
 
 // CreateMetricsReceiver creates a metrics receiver based on provided config.
 func (f *Factory) CreateMetricsReceiver(
-	logger *zap.Logger,
+	ctx context.Context,
```
[the remark this answers is not in the document]
Comment (dmitryax) [reply receiver/jaegerreceiver/factory.go]: done

File: internal/data_generator/internal/files.go [hunk 5]
```diff
@@ -33,7 +33,7 @@ const header = `// Copyright 2020 OpenTelemetry Authors
 // Code generated by "internal/data_generator/main.go". DO NOT EDIT.
 // To regenerate this file run "go run internal/data_generator/main.go".
 
-package data`
+package pdata`
```
[the remark this answers is not in the document]
Comment (dmitryax) [reply internal/data_generator/internal/files.go]: done #850

File: receiver/jaegerreceiver/trace_receiver_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (dmitryax) [reply receiver/jaegerreceiver/trace_receiver_test.go:260]: done https://github.com/open-telemetry/opentelemetry-collector/issues/851

File: translator/internaldata/oc_to_resource.go (hunk 1 above)
Comment (bogdandrutu) [reply translator/internaldata/oc_to_resource.go:46]: Separate PR then :)

Comment (dmitryax) [reply translator/internaldata/oc_to_resource.go]: done https://github.com/open-telemetry/opentelemetry-collector/pull/852

File: translator/trace/jaeger/jaegerproto_to_traces.go [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+	if parentSpanID != 0 {
+		dest.SetParentSpanID(pdata.SpanID(tracetranslator.UInt64ToByteSpanID(parentSpanID)))
+	}
+
+	attrs := jTagsToInternalAttributes(span.Tags)
+	setInternalSpanStatus(attrs, dest.Status())
+	if spanKindAttr, ok := attrs[tracetranslator.TagSpanKind]; ok {
+		dest.SetKind(jSpanKindToInternal(spanKindAttr.StringVal()))
+	}
+	dest.Attributes().InitFromMap(attrs)
```
Comment (tigrannajaryan) [inline translator/trace/jaeger/jaegerproto_to_traces.go:141]: We convert a slice to map here, then convert the map to slice again. I suspect this may be slow. Let's get a perf profile once this is merged and see what it shows. If it is slow we can probably just do a one pass and convert the source Tags slice directly to dest attributes.
This is OK for now, but let's verify it.

**Comment by dmitryax** [comment]
@tigrannajaryan 
| Can you point to the old code that does this?
Here we set ParentSpanID https://github.com/open-telemetry/opentelemetry-collector/blob/master/translator/trace/jaeger/jaegerproto_to_protospan.go#L117
And here we duplicate the already used parentSpanID reference it in SpanLinks https://github.com/open-telemetry/opentelemetry-collector/blob/master/translator/trace/jaeger/jaegerproto_to_protospan.go#L152-L178 , which looks redundant. But this is not a big concern, I'm fine with that duplication.

| You probably also need to check this with backend/UI team to make sure we don't break them if we change the behavior.
Backend reads CHILD_OF span reference, which will be recreated back from `parentSpanID` field with internal->jaeger exporter translation so should be fine.

**Comment by dmitryax** [comment]
@tigrannajaryan 
Since jaeger has only two reference types: `FOLLOWS_FROM` and `CHILD_OF` removing that duplication allows us to reconstruct Jaeger reference type on exporter: 
- any existing Span link will get `FOLLOWS_FROM` ref type
- ParentSpanId will be converted to `CHILD_OF` reference added back to the jaeger SpanRefs collection.

**Comment by dmitryax** [comment]
In addition to the comment above, looks like we are not checking for this duplicate in OC SpanLinks during OC->Jaeger translation https://github.com/open-telemetry/opentelemetry-collector/blob/75ae9198a89eb9df1d14308c89c2a4c93b6ba9e1/translator/trace/jaeger/protospan_to_jaegerproto.go#L492-L509

It means that Jaeger -> OC -> Jaeger translation most of the times adds a duplicate in Span.References for Parent span Id

**Comment by tigrannajaryan** [comment]
@dmitryax I don't know if the current parent span translation logic is intentional or a bug, I am not familiar with this part of the codebase. I am OK if you want to clean it up. We need to test the Collector fully anyway. Maybe @pjanotti knows more.

**Comment by dmitryax** [comment]
@pjanotti could you please take a look at the PR since you are the most familiar with original translation?

**Comment by dmitryax** [comment]
@pjanotti @tigrannajaryan I've applied the change to the span links translation as described above.

**Comment by dmitryax** [comment]
Appeared that HTTP Status code -> OC Status code translation is different between jaeger thrift -> OC and jaeger proto -> OC. 
- We have the http codes translation in thrift https://github.com/open-telemetry/opentelemetry-collector/blob/master/translator/trace/jaeger/jaegerthrift_to_protospan.go#L222
- And don't have it in proto translation https://github.com/open-telemetry/opentelemetry-collector/blob/master/translator/trace/jaeger/jaegerproto_to_protospan.go#L224

@owais @pjanotti do you know if it's intentional?

In this PR I've applied the translation in both thrift and proto versions.

**Comment by tigrannajaryan** [comment]
> Appeared that HTTP Status code -> OC Status code translation is different between jaeger thrift -> OC and jaeger proto -> OC.
> 
>     * We have the http codes translation in thrift https://github.com/open-telemetry/opentelemetry-collector/blob/master/translator/trace/jaeger/jaegerthrift_to_protospan.go#L222
> 
>     * And don't have it in proto translation https://github.com/open-telemetry/opentelemetry-collector/blob/master/translator/trace/jaeger/jaegerproto_to_protospan.go#L224
> 
> 
> @owais @pjanotti do you know if it's intentional?
> 
> In this PR I've applied the translation in both thrift and proto versions.

@dmitryax I know that jaeger proto version is also used by SAPM, so I would take the proto version of the translation as the canonical one and modify thirft to match it.

**Comment by tigrannajaryan** [comment]
@dmitryax actually, ignore that, I am not sure what I said is correct. Can you check with the backend team?

**Comment by tigrannajaryan** [comment]
We discussed the translation and all looks good. We will test it one more time when this goes to production.
